service/backend/pchain/indexer/parser.go

```diff
@@ -0,0 +1,318 @@
+	*blockTime = p.genesisTimestamp
+
+	var genesisTxs []*txs.Tx
+	genesisTxs = append(genesisTxs, genesisState.Validators...)
+	genesisTxs = append(genesisTxs, genesisState.Chains...)
+
+	// Genesis commit block's parent ID is the hash of genesis state
+	var genesisParentID ids.ID = hashing.ComputeHash256Array(bytes)
+
+	// Genesis Block is not indexed by the indexer, but its block ID can be accessed from block 0's parent id
+	genesisChildBlock, err := p.ParseBlockAtIndex(ctx, 1)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, utxo := range genesisState.UTXOs {
+		utxo.UTXO.Out.InitCtx(p.ctx)
+	}
+
+	genesisBlockID := genesisChildBlock.ParentID
+
+	return &ParsedGenesisBlock{
+		ParsedBlock: ParsedBlock{
+			ParentID:  genesisParentID,
+			Height:    0,
+			BlockID:   genesisBlockID,
+			BlockType: "GenesisBlock",
+			Timestamp: blockTime.UnixMilli(),
+			Txs:       genesisTxs,
+			Proposer:  Proposer{},
+		},
+		GenesisBlockData: GenesisBlockData{
+			Message:       genesisState.Message,
+			InitialSupply: genesisState.InitialSupply,
+			UTXOs:         genesisState.UTXOs,
+		},
+	}, errs.Err
+}
+
+func (p *parser) ParseCurrentBlock(ctx context.Context) (*ParsedBlock, error) {
+	err := p.initCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	height, err := p.GetPlatformHeight(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return p.ParseBlockAtIndex(ctx, height)
+}
+
+func (p *parser) ParseBlockAtIndex(ctx context.Context, index uint64) (*ParsedBlock, error) {
+	err := p.initCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	container, err := p.pChainClient.GetContainerByIndex(ctx, index-1)
+	if err != nil {
+		return nil, err
+	}
+
+	return p.parseBlockBytes(container.Bytes)
+}
+
+func (p *parser) ParseBlockWithHash(ctx context.Context, hash string) (*ParsedBlock, error) {
+	err := p.initCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	hashID, err := ids.FromString(hash)
+	if err != nil {
+		return nil, err
+	}
+
+	container, err := p.pChainClient.GetContainerByID(ctx, hashID)
+	if err != nil {
+		return nil, err
+	}
+
+	return p.parseBlockBytes(container.Bytes)
+}
+
+func (p *parser) parseBlockBytes(proposerBytes []byte) (*ParsedBlock, error) {
+	errs := wrappers.Errs{}
+
+	proposer, bytes, err := getProposerFromBytes(proposerBytes)
+	if err != nil {
+		return nil, fmt.Errorf("fetching proposer from block bytes errored with %w", err)
+	}
+
+	if p.genesisTimestamp.IsZero() {
+		return nil, errParserUninitialized
+	}
+
+	// Default block time to proposer timestamp if exists, otherwise to genesis block
+	blockTimestamp := new(time.Time)
+
+	if proposer.Timestamp > genesisTimestamp {
+		*blockTimestamp = time.Unix(proposer.Timestamp, 0)
+	} else {
+		*blockTimestamp = time.Unix(genesisTimestamp, 0)
+	}
+
+	var blk blocks.Block
+	ver, err := p.codec.Unmarshal(bytes, &blk)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshaling block bytes errored with %w", err)
+	}
+	blkID := ids.ID(hashing.ComputeHash256Array(bytes))
```

Comment:
consider `blkID := blk.ID()` here. In general it'd be best to pull IDs out of blocks rather than recalcutiong them here. This should simplify thing should ID implementation ever change